File: AgentBI/src/api/capability_settings.py
from __future__ import annotations

from typing import Any

from fastapi import APIRouter, HTTPException, Query, Request
from pydantic import ValidationError

from AgentBI.src.agents.assistant_registry import CAPABILITIES
from AgentBI.src.api.dependencies import get_chat_repository
from AgentBI.src.schemas.capability_settings_schema import (
    CAPABILITY_CONFIG_FIELDS,
    CapabilitySettingsResponse,
    CapabilitySettingsUpdate,
    resolve_capability_config,
)

router = APIRouter(tags=["capability-settings"])


def validate_capability_config(capability_id: str, config: dict[str, Any]) -> dict[str, Any]:
    if capability_id not in CAPABILITIES:
        raise HTTPException(status_code=404, detail="能力不存在")
    try:
        return resolve_capability_config(capability_id, config)
    except ValidationError as error:
        raise HTTPException(status_code=422, detail=error.errors()) from error


def _response(repository: Any, user_id: str, capability_id: str) -> CapabilitySettingsResponse:
    capability = CAPABILITIES.get(capability_id)
    if not capability:
        raise HTTPException(status_code=404, detail="能力不存在")
    config = validate_capability_config(
        capability_id,
        repository.get_capability_config(user_id, capability_id) or {},
    )
    return CapabilitySettingsResponse(
        capability_id=capability_id,
        display_name=capability["name"],
        description=capability["description"],
        kind=capability["kind"],
        config=config,
        fields=CAPABILITY_CONFIG_FIELDS.get(capability_id, []),
    )


@router.get("/capabilities/settings", response_model=list[CapabilitySettingsResponse])
def list_capability_settings(request: Request, user_id: str = Query(min_length=1)):
    repository = get_chat_repository(request)
    return [_response(repository, user_id, capability_id) for capability_id in CAPABILITIES]


@router.put("/capabilities/{capability_id}/settings", response_model=CapabilitySettingsResponse)
def save_capability_settings(
    capability_id: str,
    request: Request,
    payload: CapabilitySettingsUpdate,
    user_id: str = Query(min_length=1),
):
    repository = get_chat_repository(request)
    config = validate_capability_config(capability_id, payload.config)
    repository.save_capability_config(user_id, capability_id, config)
    return _response(repository, user_id, capability_id)

File: Agent-vue/src/api/capability-settings.ts
import type { CapabilitySettings } from './chat-types'

const BASE = '/api'

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await fetch(`${BASE}${path}`, init)
  if (!response.ok)
    throw new Error(
      (await response.json().catch(() => null))?.detail || `能力配置请求失败 (${response.status})`,
    )
  return response.json() as Promise<T>
}

export const listCapabilitySettings = (userId: string) =>
  request<CapabilitySettings[]>(`/capabilities/settings?user_id=${encodeURIComponent(userId)}`)

export const saveCapabilitySettings = (
  userId: string,
  capabilityId: string,
  config: CapabilitySettings['config'],
) =>
  request<CapabilitySettings>(
    `/capabilities/${encodeURIComponent(capabilityId)}/settings?user_id=${encodeURIComponent(userId)}`,
    {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ config }),
    },
  )
